ads/google/a4a/line-delimited-response-handler.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ import {dev} from '../../../src/log';
+ import {isObject} from '../../../src/types';
+
+ /**
+  * Creates an XHR streaming request with expected response format of repeated
+  * pairs of JSON object metadata then string HTML delimited by line returns.
+  *
+  * @param {!../../../src/service/xhr-impl.Xhr} xhr
+  * @param {string} url
+  * @param {!function(string, !Object<string, *>, boolean)} slotCallback called
+  *    with creative, metadata, and boolean indicating if completed.  Failure to
+  *    JSON parse metadata will cause empty object to be given.
+  * @param {?../../../src/service/xhr-impl.FetchInitDef=} opt_init
+  * @return {!Promise<!../../../src/service/xhr-impl.FetchResponse>} response,
+  *    note that accessing body will cause error as its consumed in streaming.
+  */
+ export function fetchLineDelimitedChunks(xhr, url, slotCallback, opt_init) {
+   return xhr.fetch(url, opt_init)
+     .then(response => {
+       let metadata;
+       let snippet = '';
+       const chunkCallback = (chunk, done) => {
+         const regex = /([^\n]*)(\n)?/g;
+         let match;
+         while ((match = regex.exec(chunk))) {
+           snippet += match[1];
+           if (match[2]) {
+             if (metadata) {
```

Comment:
Yes, after refactor suggested by @jridgewell I believe this is more clear given it now utilizes a function named metaJsonCreativeGrouper